From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Tue, 18 Jul 2023 14:47:02 -0700
Subject: [PATCH] Add Lifecycle Event system

This event system is separate from Bukkit's event system and is
meant for managing resources across reloads and from points in the
PluginBootstrap.

diff --git a/src/main/java/io/papermc/paper/plugin/bootstrap/BootstrapContext.java b/src/main/java/io/papermc/paper/plugin/bootstrap/BootstrapContext.java
index 08f2050356acaf74e3210416760e3873c2dafd2c..08525c498cab1268d6e001cc703d9eb562872a96 100644
--- a/src/main/java/io/papermc/paper/plugin/bootstrap/BootstrapContext.java
+++ b/src/main/java/io/papermc/paper/plugin/bootstrap/BootstrapContext.java
@@ -1,5 +1,6 @@
 package io.papermc.paper.plugin.bootstrap;
 
+import io.papermc.paper.plugin.lifecycle.event.LifecycleEventOwner;
 import org.jetbrains.annotations.ApiStatus;
 
 /**
@@ -10,5 +11,5 @@ import org.jetbrains.annotations.ApiStatus;
  */
 @ApiStatus.Experimental
 @ApiStatus.NonExtendable
-public interface BootstrapContext extends PluginProviderContext {
+public interface BootstrapContext extends PluginProviderContext, LifecycleEventOwner<BootstrapContext> {
 }
diff --git a/src/main/java/io/papermc/paper/plugin/lifecycle/LifecycleEvent.java b/src/main/java/io/papermc/paper/plugin/lifecycle/LifecycleEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..5619e10bd6262aa766ebb855f8656c9c85bff025
--- /dev/null
+++ b/src/main/java/io/papermc/paper/plugin/lifecycle/LifecycleEvent.java
@@ -0,0 +1,8 @@
+package io.papermc.paper.plugin.lifecycle;
+
+import org.jetbrains.annotations.ApiStatus;
+
+@ApiStatus.Experimental
+@ApiStatus.NonExtendable
+public interface LifecycleEvent {
+}
diff --git a/src/main/java/io/papermc/paper/plugin/lifecycle/dummy/DummyResourceRegistrar.java b/src/main/java/io/papermc/paper/plugin/lifecycle/dummy/DummyResourceRegistrar.java
new file mode 100644
index 0000000000000000000000000000000000000000..359bedf119d44e1a7f80d6edc16fe5f0a5afc2ac
--- /dev/null
+++ b/src/main/java/io/papermc/paper/plugin/lifecycle/dummy/DummyResourceRegistrar.java
@@ -0,0 +1,6 @@
+package io.papermc.paper.plugin.lifecycle.dummy;
+
+import io.papermc.paper.plugin.lifecycle.event.registrar.ResourceRegistrar;
+
+public class DummyResourceRegistrar implements ResourceRegistrar {
+}
diff --git a/src/main/java/io/papermc/paper/plugin/lifecycle/dummy/NonRegistrarEvent.java b/src/main/java/io/papermc/paper/plugin/lifecycle/dummy/NonRegistrarEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..5d18c53467f4c5277c5fc245aa2ce0d1ad827617
--- /dev/null
+++ b/src/main/java/io/papermc/paper/plugin/lifecycle/dummy/NonRegistrarEvent.java
@@ -0,0 +1,10 @@
+package io.papermc.paper.plugin.lifecycle.dummy;
+
+import io.papermc.paper.plugin.lifecycle.LifecycleEvent;
+
+public class NonRegistrarEvent implements LifecycleEvent {
+
+    public void someNonRegistrarRelatedThing() {
+        System.out.println("non registrar event");
+    }
+}
diff --git a/src/main/java/io/papermc/paper/plugin/lifecycle/event/LifecycleEventHandler.java b/src/main/java/io/papermc/paper/plugin/lifecycle/event/LifecycleEventHandler.java
new file mode 100644
index 0000000000000000000000000000000000000000..170b8048fb353ffbad4fb87e4b38bd75079137e8
--- /dev/null
+++ b/src/main/java/io/papermc/paper/plugin/lifecycle/event/LifecycleEventHandler.java
@@ -0,0 +1,10 @@
+package io.papermc.paper.plugin.lifecycle.event;
+
+import io.papermc.paper.plugin.lifecycle.LifecycleEvent;
+import org.jetbrains.annotations.NotNull;
+
+@FunctionalInterface
+public interface LifecycleEventHandler<E extends LifecycleEvent> {
+
+    void run(@NotNull E event);
+}
diff --git a/src/main/java/io/papermc/paper/plugin/lifecycle/event/LifecycleEventOwner.java b/src/main/java/io/papermc/paper/plugin/lifecycle/event/LifecycleEventOwner.java
new file mode 100644
index 0000000000000000000000000000000000000000..800d28ad2d01068085effa2753263ef246ec7515
--- /dev/null
+++ b/src/main/java/io/papermc/paper/plugin/lifecycle/event/LifecycleEventOwner.java
@@ -0,0 +1,27 @@
+package io.papermc.paper.plugin.lifecycle.event;
+
+import io.papermc.paper.plugin.lifecycle.LifecycleEvent;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+@ApiStatus.Experimental
+@ApiStatus.NonExtendable
+public interface LifecycleEventOwner<O extends LifecycleEventOwner<O>> {
+
+    @SuppressWarnings("unchecked")
+    private O self() {
+        return (O) this;
+    }
+
+    default <E extends LifecycleEvent, T extends LifecycleEventType<? super O, ? super E>> void handleLifecycleEvent(final @NotNull T lifecycleEventType, final @NotNull LifecycleEventHandler<? super E> lifecycleEventHandler) {
+        LifecycleEventTypeProvider.PROVIDER.register(this.self(), lifecycleEventType, lifecycleEventHandler);
+    }
+
+    default <E extends LifecycleEvent, T extends LifecycleEventType<? super O, ? super E>> void handleLifecycleEventAsMonitor(final @NotNull T lifecycleEventType, final @NotNull LifecycleEventHandler<? super E> lifecycleEventHandler) {
+        LifecycleEventTypeProvider.PROVIDER.registerAsMonitor(this.self(), lifecycleEventType, lifecycleEventHandler);
+    }
+
+    default <E extends LifecycleEvent, T extends LifecycleEventType.Prioritized<? super O, ? super E>> void handleLifecycleEventWithPriority(final @NotNull T lifecycleEventType, final @NotNull LifecycleEventHandler<? super E> lifecycleEventHandler, int priority) {
+        LifecycleEventTypeProvider.PROVIDER.registerWithPriority(this.self(), lifecycleEventType, lifecycleEventHandler, priority);
+    }
+}
diff --git a/src/main/java/io/papermc/paper/plugin/lifecycle/event/LifecycleEventType.java b/src/main/java/io/papermc/paper/plugin/lifecycle/event/LifecycleEventType.java
new file mode 100644
index 0000000000000000000000000000000000000000..a06c18bfe066f84462a74e0b6bfc29650bfdf757
--- /dev/null
+++ b/src/main/java/io/papermc/paper/plugin/lifecycle/event/LifecycleEventType.java
@@ -0,0 +1,19 @@
+package io.papermc.paper.plugin.lifecycle.event;
+
+import io.papermc.paper.plugin.lifecycle.LifecycleEvent;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+@SuppressWarnings("unused")
+@ApiStatus.Experimental
+@ApiStatus.NonExtendable
+public interface LifecycleEventType<O extends LifecycleEventOwner<O>, E extends LifecycleEvent> {
+
+    @NotNull String name();
+
+    @ApiStatus.Experimental
+    @ApiStatus.NonExtendable
+    interface Prioritized<O extends LifecycleEventOwner<O>, E extends LifecycleEvent> extends LifecycleEventType<O, E> {
+
+    }
+}
diff --git a/src/main/java/io/papermc/paper/plugin/lifecycle/event/LifecycleEventTypeProvider.java b/src/main/java/io/papermc/paper/plugin/lifecycle/event/LifecycleEventTypeProvider.java
new file mode 100644
index 0000000000000000000000000000000000000000..6b19ce7ab19133269406a9ca6cdf49c1859308d9
--- /dev/null
+++ b/src/main/java/io/papermc/paper/plugin/lifecycle/event/LifecycleEventTypeProvider.java
@@ -0,0 +1,25 @@
+package io.papermc.paper.plugin.lifecycle.event;
+
+import io.papermc.paper.plugin.configuration.PluginMeta;
+import io.papermc.paper.plugin.lifecycle.LifecycleEvent;
+import java.util.ServiceLoader;
+import java.util.function.Function;
+import org.jetbrains.annotations.ApiStatus;
+
+@ApiStatus.Internal
+interface LifecycleEventTypeProvider {
+
+    LifecycleEventTypeProvider PROVIDER = ServiceLoader.load(LifecycleEventTypeProvider.class)
+        .findFirst()
+        .orElseThrow();
+
+    <O extends LifecycleEventOwner<O>, E extends LifecycleEvent, T extends LifecycleEventType<? super O, ? super E>> void register(O owner, T eventType, LifecycleEventHandler<? super E> lifecycleEventHandler);
+
+    <O extends LifecycleEventOwner<O>, E extends LifecycleEvent, T extends LifecycleEventType<? super O, ? super E>> void registerAsMonitor(O owner, T eventType, LifecycleEventHandler<? super E> lifecycleEventHandler);
+
+    <O extends LifecycleEventOwner<O>, E extends LifecycleEvent, T extends LifecycleEventType.Prioritized<? super O, ? super E>> void registerWithPriority(O owner, T eventType, LifecycleEventHandler<? super E> lifecycleEventHandler, int priority);
+
+    <O extends LifecycleEventOwner<O>, E extends LifecycleEvent> LifecycleEventType<O, E> simple(String name, Function<? super O, ? extends PluginMeta> getPluginMeta);
+
+    <O extends LifecycleEventOwner<O>, E extends LifecycleEvent> LifecycleEventType.Prioritized<O, E> prioritized(String name, Function<? super O, ? extends PluginMeta> getPluginMeta);
+}
diff --git a/src/main/java/io/papermc/paper/plugin/lifecycle/event/LifecycleEvents.java b/src/main/java/io/papermc/paper/plugin/lifecycle/event/LifecycleEvents.java
new file mode 100644
index 0000000000000000000000000000000000000000..b8a5e77e14b27dfdf558ad3a090dc6a0ac88f3ee
--- /dev/null
+++ b/src/main/java/io/papermc/paper/plugin/lifecycle/event/LifecycleEvents.java
@@ -0,0 +1,60 @@
+package io.papermc.paper.plugin.lifecycle.event;
+
+import io.papermc.paper.plugin.bootstrap.BootstrapContext;
+import io.papermc.paper.plugin.configuration.PluginMeta;
+import io.papermc.paper.plugin.lifecycle.LifecycleEvent;
+import io.papermc.paper.plugin.lifecycle.event.registrar.RegistrarEvent;
+import io.papermc.paper.plugin.lifecycle.dummy.DummyResourceRegistrar;
+import io.papermc.paper.plugin.lifecycle.dummy.NonRegistrarEvent;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Function;
+import org.bukkit.plugin.Plugin;
+import org.jetbrains.annotations.ApiStatus;
+
+@ApiStatus.Experimental
+public final class LifecycleEvents {
+
+    static final List<LifecycleEventType<?, ?>> LIFECYCLE_EVENT_TYPES = new ArrayList<>();
+
+    public static final LifecycleEventType<BootstrapContext, RegistrarEvent.Reloadable<DummyResourceRegistrar>> DUMMY = bootstrap("dummy");
+    public static final LifecycleEventType.Prioritized<Plugin, RegistrarEvent<DummyResourceRegistrar>> DUMMY_STATIC = pluginPrioritized("dummy_static");
+    public static final LifecycleEventType<BootstrapContext, NonRegistrarEvent> NON_REGISTRAR_RELATED_EVENT = bootstrap("non_registrar");
+
+    @ApiStatus.Internal
+    private static <E extends LifecycleEvent> LifecycleEventType<Plugin, E> plugin(final String name) {
+        return createSimple(name, Plugin::getPluginMeta);
+    }
+
+    @ApiStatus.Internal
+    private static <E extends LifecycleEvent> LifecycleEventType.Prioritized<Plugin, E> pluginPrioritized(final String name) {
+        return createPrioritized(name, Plugin::getPluginMeta);
+    }
+
+    @ApiStatus.Internal
+    private static <E extends LifecycleEvent> LifecycleEventType<BootstrapContext, E> bootstrap(final String name) {
+        return createSimple(name, BootstrapContext::getConfiguration);
+    }
+
+    @ApiStatus.Internal
+    private static <E extends LifecycleEvent> LifecycleEventType.Prioritized<BootstrapContext, E> bootstrapPrioritized(final String name) {
+        return createPrioritized(name, BootstrapContext::getConfiguration);
+    }
+
+    @ApiStatus.Internal
+    private static <O extends LifecycleEventOwner<O>, E extends LifecycleEvent> LifecycleEventType<O, E> createSimple(final String name, final Function<O, ? extends PluginMeta> getPluginMeta) {
+        final LifecycleEventType<O, E> lifecycleEventType = LifecycleEventTypeProvider.PROVIDER.simple(name, getPluginMeta);
+        LIFECYCLE_EVENT_TYPES.add(lifecycleEventType);
+        return lifecycleEventType;
+    }
+
+    @ApiStatus.Internal
+    private static <O extends LifecycleEventOwner<O>, E extends LifecycleEvent> LifecycleEventType.Prioritized<O, E> createPrioritized(final String name, final Function<O, ? extends PluginMeta> getPluginMeta) {
+        final LifecycleEventType.Prioritized<O, E> lifecycleEventType = LifecycleEventTypeProvider.PROVIDER.prioritized(name, getPluginMeta);
+        LIFECYCLE_EVENT_TYPES.add(lifecycleEventType);
+        return lifecycleEventType;
+    }
+
+    private LifecycleEvents() {
+    }
+}
diff --git a/src/main/java/io/papermc/paper/plugin/lifecycle/event/registrar/RegistrarEvent.java b/src/main/java/io/papermc/paper/plugin/lifecycle/event/registrar/RegistrarEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..679974e4680d75c59077862fcbbf3867a1fb0989
--- /dev/null
+++ b/src/main/java/io/papermc/paper/plugin/lifecycle/event/registrar/RegistrarEvent.java
@@ -0,0 +1,25 @@
+package io.papermc.paper.plugin.lifecycle.event.registrar;
+
+import io.papermc.paper.plugin.lifecycle.LifecycleEvent;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+@ApiStatus.Experimental
+@ApiStatus.NonExtendable
+public interface RegistrarEvent<R extends ResourceRegistrar> extends LifecycleEvent {
+
+    @NotNull R registrar();
+
+    @ApiStatus.Experimental
+    @ApiStatus.NonExtendable
+    interface Reloadable<R extends ResourceRegistrar> extends RegistrarEvent<R> {
+
+        @NotNull Cause cause();
+
+        @ApiStatus.Experimental
+        enum Cause {
+            INITIAL,
+            RELOAD
+        }
+    }
+}
diff --git a/src/main/java/io/papermc/paper/plugin/lifecycle/event/registrar/ResourceRegistrar.java b/src/main/java/io/papermc/paper/plugin/lifecycle/event/registrar/ResourceRegistrar.java
new file mode 100644
index 0000000000000000000000000000000000000000..6bebcd6e61b98f218aa464b1005c7daa21f8e2b6
--- /dev/null
+++ b/src/main/java/io/papermc/paper/plugin/lifecycle/event/registrar/ResourceRegistrar.java
@@ -0,0 +1,12 @@
+package io.papermc.paper.plugin.lifecycle.event.registrar;
+
+import org.jetbrains.annotations.ApiStatus;
+
+/**
+ * To be implemented by types that provide ways to register types
+ * either on server start or during a reload
+ */
+@ApiStatus.Experimental
+@ApiStatus.NonExtendable
+public interface ResourceRegistrar {
+}
diff --git a/src/main/java/org/bukkit/plugin/Plugin.java b/src/main/java/org/bukkit/plugin/Plugin.java
index 4eb639fbb46a0848be207149ea433455550fae1c..178b58b9d593384e3cf74ec1630f651d21aeed7f 100644
--- a/src/main/java/org/bukkit/plugin/Plugin.java
+++ b/src/main/java/org/bukkit/plugin/Plugin.java
@@ -1,5 +1,6 @@
 package org.bukkit.plugin;
 
+import io.papermc.paper.plugin.lifecycle.event.LifecycleEventOwner;
 import java.io.File;
 import java.io.InputStream;
 import java.util.logging.Logger;
@@ -16,7 +17,7 @@ import org.jetbrains.annotations.Nullable;
  * <p>
  * The use of {@link PluginBase} is recommended for actual Implementation
  */
-public interface Plugin extends TabExecutor {
+public interface Plugin extends TabExecutor, LifecycleEventOwner<Plugin> { // Paper
     /**
      * Returns the folder that the plugin data's files are located in. The
      * folder may not yet exist.
diff --git a/src/main/java/org/bukkit/plugin/java/JavaPlugin.java b/src/main/java/org/bukkit/plugin/java/JavaPlugin.java
index 5cd236965de12392d8c7aa81307c0ff1cc8673b1..90f1139508ad98484c851c116a00c09f81663486 100644
--- a/src/main/java/org/bukkit/plugin/java/JavaPlugin.java
+++ b/src/main/java/org/bukkit/plugin/java/JavaPlugin.java
@@ -278,7 +278,14 @@ public abstract class JavaPlugin extends PluginBase {
             isEnabled = enabled;
 
             if (isEnabled) {
+                try { // Paper
+                this.allowLifecycleEventHandlerRegistration = true; // Paper - only allow registration during onEnable
                 onEnable();
+                // Paper start - only allow registration during onEnable
+                } finally {
+                    this.allowLifecycleEventHandlerRegistration = false;
+                }
+                // Paper end
             } else {
                 onDisable();
             }
@@ -456,4 +463,25 @@ public abstract class JavaPlugin extends PluginBase {
         }
         return plugin;
     }
+
+    // Paper start - lifecycle events
+    private boolean allowLifecycleEventHandlerRegistration = true;
+    @Override
+    public <E extends io.papermc.paper.plugin.lifecycle.LifecycleEvent, T extends io.papermc.paper.plugin.lifecycle.event.LifecycleEventType<? super org.bukkit.plugin.Plugin, ? super E>> void handleLifecycleEvent(final @NotNull T lifecycleEventType, final io.papermc.paper.plugin.lifecycle.event.@NotNull LifecycleEventHandler<? super E> lifecycleEventHandler) {
+        Preconditions.checkState(this.allowLifecycleEventHandlerRegistration, "Cannot add lifecycle event handlers after onEnable");
+        super.handleLifecycleEvent(lifecycleEventType, lifecycleEventHandler);
+    }
+
+    @Override
+    public <E extends io.papermc.paper.plugin.lifecycle.LifecycleEvent, T extends io.papermc.paper.plugin.lifecycle.event.LifecycleEventType<? super org.bukkit.plugin.Plugin, ? super E>> void handleLifecycleEventAsMonitor(final @NotNull T lifecycleEventType, final io.papermc.paper.plugin.lifecycle.event.@NotNull LifecycleEventHandler<? super E> lifecycleEventHandler) {
+        Preconditions.checkState(this.allowLifecycleEventHandlerRegistration, "Cannot add lifecycle event handlers after onEnable");
+        super.handleLifecycleEventAsMonitor(lifecycleEventType, lifecycleEventHandler);
+    }
+
+    @Override
+    public <E extends io.papermc.paper.plugin.lifecycle.LifecycleEvent, T extends io.papermc.paper.plugin.lifecycle.event.LifecycleEventType.Prioritized<? super org.bukkit.plugin.Plugin, ? super E>> void handleLifecycleEventWithPriority(final @NotNull T lifecycleEventType, final io.papermc.paper.plugin.lifecycle.event.@NotNull LifecycleEventHandler<? super E> lifecycleEventHandler, final int priority) {
+        Preconditions.checkState(this.allowLifecycleEventHandlerRegistration, "Cannot add lifecycle event handlers after onEnable");
+        super.handleLifecycleEventWithPriority(lifecycleEventType, lifecycleEventHandler, priority);
+    }
+    // Paper end - lifecycle events
 }
